File: zoom-climb/src/lib/dsl/registry.ts
/**
 * The DSL registry — validator adapters plus per-DSL metadata.
 *
 * Every Buhera module carries its own DSL with its own compiler, and those
 * compilers disagree on their return shape: some return {valid, errors},
 * some {ok, errors}, some throw. Each adapter here normalizes one of them
 * to a single contract so the generate -> validate -> repair loop stays
 * DSL-agnostic:
 *
 *     validate(source) -> { ok, errors: [{ message, line? }] }
 *
 * The validator is the ground truth. This module ships no facts about what
 * good code looks like — it proposes code and lets the DSL's real compiler
 * judge. A generated chunk is valid iff the module's own compiler accepts it.
 *
 * Adding a DSL is one entry plus its normalizer. Nothing in the generator
 * changes.
 */

import { parseVahera } from "./vahera";
import type { DslEntry, ValidationResult } from "./types";

/**
 * Pull a 1-based line number out of a thrown parser message of the form
 * "line N: ...". Returns null when the message carries no line marker.
 */
function lineFromMessage(message: string): number | null {
  const m = /(?:^|\b)line\s+(\d+)\b/i.exec(message || "");
  return m ? parseInt(m[1], 10) : null;
}

/**
 * vaHera — `parseVahera` THROWS on the first invalid line, embedding
 * "line N:" in the message. Parsing is pure (no kernel needed), so this is
 * a safe check without any execution.
 */
export function validateVahera(source: string): ValidationResult {
  try {
    parseVahera(source);
    return { ok: true, errors: [] };
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return { ok: false, errors: [{ message, line: lineFromMessage(message) }] };
  }
}

/**
 * The registry.
 *
 * vaHera is first because it is the one DSL whose compiler is pure and
 * portable. Turbulance (graffiti/lang/parser.ts), shapeshifter
 * (lavoisier/shapeshifter/compiler.js), SCOPE and purpose are added here as
 * their compilers are wired in — each is one entry plus a normalizer.
 *
 * `acceptsFragment` is not cosmetic: vaHera is line-oriented with no program
 * structure, so a chunk parses exactly as a script does. A DSL with block
 * structure or accumulating REPL semantics will not have that property, and
 * chunk-extent generation for it needs a different validation call.
 */
export const DSL_REGISTRY: Record<string, DslEntry> = {
  vahera: {
    id: "vahera",
    label: "vaHera",
    validate: validateVahera,
    moduleId: "vahera",
    packId: "vahera",
    acceptsFragment: true,
  },
};

export function listDsls(): DslEntry[] {
  return Object.values(DSL_REGISTRY);
}

export function getDsl(dslId: string): DslEntry | null {
  return Object.prototype.hasOwnProperty.call(DSL_REGISTRY, dslId)
    ? DSL_REGISTRY[dslId]
    : null;
}

/**
 * Validate against the named DSL's real compiler. Throws on an unknown DSL
 * id — that is a programming error, categorically different from invalid
 * source, which returns {ok:false}.
 */
export function validate(dslId: string, source: string): ValidationResult {
  const dsl = getDsl(dslId);
  if (!dsl) throw new Error(`unknown DSL: "${dslId}"`);
  return dsl.validate(source);
}

File: crates/zangalewa-dsl/src/registry.rs
//! The DSL registry — which languages Zangalewa can write, and whose compiler
//! judges each one.
//!
//! `validate` is the module's REAL compiler, normalized to a common verdict
//! shape. Nothing here approximates a grammar or second-guesses a compiler:
//! the whole design rests on asking the owning module whether the code parses,
//! because that is the only judgement this module is entitled to make.
//!
//! Adding a DSL is one entry plus a normalizer. Mirrors
//! `zoom-climb/src/lib/dsl/registry.ts`.

use crate::types::{ValidationError, ValidationResult};
use crate::vahera::parse_vahera;

/// One DSL the generator can target.
pub struct DslEntry {
    pub id: &'static str,
    pub label: &'static str,
    /// Who executes the code once it reaches the OS. Recorded here, never
    /// acted on by this module.
    pub module_id: &'static str,
    /// Names the grounding material.
    pub pack_id: &'static str,
    /// Whether this DSL's compiler accepts a fragment as well as a whole
    /// program. Line-oriented DSLs (vaHera) do; DSLs with program structure
    /// or accumulating REPL semantics (SCOPE) may not, in which case chunk
    /// extent needs a different validation call than script extent.
    pub accepts_fragment: bool,
    pub validate: fn(&str) -> ValidationResult,
}

pub fn validate_vahera(source: &str) -> ValidationResult {
    match parse_vahera(source) {
        Ok(_) => ValidationResult::accepted(),
        Err(e) => ValidationResult::rejected(vec![ValidationError {
            message: e.message,
            line: Some(e.line),
        }]),
    }
}

static DSLS: &[DslEntry] = &[DslEntry {
    id: "vahera",
    label: "vaHera",
    module_id: "vahera",
    pack_id: "vahera",
    accepts_fragment: true,
    validate: validate_vahera,
}];

pub fn get_dsl(id: &str) -> Option<&'static DslEntry> {
    DSLS.iter().find(|d| d.id == id)
}

pub fn list_dsls() -> &'static [DslEntry] {
    DSLS
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn accepts_valid_source() {
        let r = validate_vahera("memory list\ndemon sort");
        assert!(r.ok);
        assert!(r.errors.is_empty());
    }

    /// The line anchor is what the repair prompt points the model at. Losing
    /// it turns a targeted correction into a guess.
    #[test]
    fn rejection_carries_the_line_anchor() {
        let r = validate_vahera("memory list\nmemory store \"a\" == \"b\"");
        assert!(!r.ok);
        assert_eq!(r.errors[0].line, Some(2));
        assert!(r.errors[0].message.contains("unknown vaHera"));
    }

    #[test]
    fn vahera_is_registered_and_accepts_fragments() {
        let dsl = get_dsl("vahera").expect("vahera must be registered");
        assert!(dsl.accepts_fragment);
        assert!(get_dsl("nope").is_none());
    }
}
